### PR TITLE
[SPARK-45558][SS] Introduce a metadata file for streaming stateful operator

### PR DESCRIPTION
**Pull request by chaoqin-li1123**
### What changes were proposed in this pull request?
Introduce a metadata file for streaming stateful operator, write metadata for stateful operator during planning.
The information to store in the metadata file:
- operator name (no need to be unique among stateful operators in the query)
- state store name
- numColumnsPrefixKey: > 0 if prefix scan is enabled, 0 otherwise
The body of metadata file will be in json format. The metadata file will be versioned just as other streaming metadata file to be future proof.

### Why are the changes needed?
The metadata file will improve expose more information about the state store, improves debugability and facilitate the development of state related feature such as reading and writing state and state repartitioning.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Add unit test and integration tests

### Was this patch authored or co-authored using generative AI tooling?
No

### CODE REVIEW

**Comment by chaoqin-li1123** [comment]
@HeartSaVioR

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/OperatorStateMetadata.scala
Comment (HeartSaVioR) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/OperatorStateMetadata.scala:1]: General comment: we'd like to have a class doc for each trait/class.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/OperatorStateMetadataSuite.scala [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+    val statePath = new Path(checkpointDir.toString, "state/0")
+    val operatorMetadata = new OperatorStateMetadataReader(statePath, hadoopConf).read()
+      .asInstanceOf[OperatorStateMetadataV1]
+
+    val expectedMetadata = OperatorStateMetadataV1(
+      OperatorInfoV1(0, "sessionWindowStateStoreSaveExec"),
+      Array(StateStoreMetadataV1("default", 1, spark.sessionState.conf.numShufflePartitions))
+    )
+    assert(sameOperatorStateMetadata(operatorMetadata, expectedMetadata))
+  }
```
Comment (HeartSaVioR) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/OperatorStateMetadataSuite.scala:157]: Let's add another test to verify multiple stateful operators (opId being 0, 1, 2, etc.).

**Review by HeartSaVioR — COMMENTED** [review]
Looks good in overall, left comments. Thanks for working on this!

File: sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/OperatorStateMetadataSuite.scala (hunk 1 above)
Comment (chaoqin-li1123) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/OperatorStateMetadataSuite.scala:177]: Multiple operators test added.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/OperatorStateMetadata.scala
Comment (chaoqin-li1123) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/OperatorStateMetadata.scala:1]: class doc added/

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/IncrementalExecution.scala [hunk 2]
```diff
@@ -177,6 +183,20 @@ class IncrementalExecution(
...
+    override val rule: PartialFunction[SparkPlan, SparkPlan] = {
+      case stateStoreWriter: StateStoreWriter =>
+        if (isFirstBatch) {
+          val metadata = stateStoreWriter.operatorStateMetadata()
+          val metadataWriter = new OperatorStateMetadataWriter(new Path(
+            checkpointLocation, stateStoreWriter.getStateInfo.operatorId.toString), hadoopConf)
+          metadataWriter.write(metadata)
+        }
+        stateStoreWriter
+      case plan: SparkPlan => plan
```
[the remark this answers is not in the document]
Comment (chaoqin-li1123) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/IncrementalExecution.scala]: Thanks for reminding, we can simplify the code a lot.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/OperatorStateMetadataSuite.scala [hunk 3]
```diff
@@ -32,30 +32,37 @@ class OperatorStateMetadataSuite extends StreamTest with SharedSparkSession {
...
+    checkpointDir = Utils.createTempDir()
+  }
+
+  private def checkOperatorStateMetadata(
+      operatorId: Int,
+      expectedMetadata: OperatorStateMetadataV1): Unit = {
+    val statePath = new Path(checkpointDir.toString, s"state/$operatorId")
+    val operatorMetadata = new OperatorStateMetadataReader(statePath, hadoopConf).read()
+      .asInstanceOf[OperatorStateMetadataV1]
+    // println("doodoo: " + operatorMetadata.stateStoreInfo)
```
Comment (HeartSaVioR) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/OperatorStateMetadataSuite.scala:50]: nit: should be removed

File: sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/OperatorStateMetadataSuite.scala [hunk 4]
```diff
@@ -32,30 +32,37 @@ class OperatorStateMetadataSuite extends StreamTest with SharedSparkSession {
 
   private lazy val hadoopConf = spark.sessionState.newHadoopConf()
 
+  private var checkpointDir = Utils.createTempDir()
```
Comment (HeartSaVioR) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/OperatorStateMetadataSuite.scala:35]: Our best practice is to use withTempDir per each test. Could we please follow the pattern and remove this and beforeEach?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/OperatorStateMetadataSuite.scala [hunk 5]
```diff
@@ -145,14 +142,37 @@ class OperatorStateMetadataSuite extends StreamTest with SharedSparkSession {
...
-
     val expectedMetadata = OperatorStateMetadataV1(
       OperatorInfoV1(0, "sessionWindowStateStoreSaveExec"),
       Array(StateStoreMetadataV1("default", 1, spark.sessionState.conf.numShufflePartitions))
     )
-    assert(sameOperatorStateMetadata(operatorMetadata, expectedMetadata))
+    checkOperatorStateMetadata(0, expectedMetadata)
+  }
+
+  test("Stateful operator metadata for multiple operators.") {
```
Comment (HeartSaVioR) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/OperatorStateMetadataSuite.scala:152]: nit: remove `.` at the end for consistency

**Review by HeartSaVioR — COMMENTED** [review]
Only minors and nits.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/OperatorStateMetadataSuite.scala (hunk 3 above)
Comment (chaoqin-li1123) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/OperatorStateMetadataSuite.scala]: Sorry, forgot to delete the debug print.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/OperatorStateMetadataSuite.scala (hunk 4 above)
Comment (chaoqin-li1123) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/OperatorStateMetadataSuite.scala]: Refactored.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/OperatorStateMetadataSuite.scala (hunk 5 above)
Comment (chaoqin-li1123) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/OperatorStateMetadataSuite.scala]: Removed.

**Review by HeartSaVioR — APPROVED** [review]
+1 pending CI. Thanks for the work!

**Comment by HeartSaVioR** [comment]
Thanks! Merging to master.

**Pull request closed without merging**